File: oop/06-inheritance-vs-composition/typescript/inheritance/savingsAccount.ts
/*
 * Demonstrates: Inheritance vs Composition - Inheritance Approach
 * Concept: SavingsAccount inherits interest behaviour from InterestBearingAccount
 */

import { InterestBearingAccount } from "./interestBearingAccount";

export class SavingsAccount extends InterestBearingAccount {
    public readonly minimumBalance: number;

    constructor(
        accountNumber: string,
        accountHolder: string,
        openingBalance: number,
        interestRate: number,
        minimumBalance: number
    ) {
        super(accountNumber, accountHolder, openingBalance, interestRate);
        this.minimumBalance = minimumBalance;
    }

    override withdraw(amount: number): void {
        if (amount <= 0) {
            throw new Error("Amount must be greater than zero.");
        }

        const resultingBalance = this._balance - amount;

        if (resultingBalance < this.minimumBalance) {
            throw new Error(`Cannot withdraw. Minimum balance of R${this.minimumBalance.toFixed(2)} required.`);
        }

        this._balance = resultingBalance;
    }
}

File: oop/06-inheritance-vs-composition/csharp/Inheritance/SavingsAccount.cs
/*
 * Demonstrates: Inheritance vs Composition - Inheritance Approach
 * Concept: SavingsAccount inherits interest behaviour from InterestBearingAccount
 */

namespace FinanceDomain.InheritanceVsComposition.Inheritance;

public class SavingsAccount : InterestBearingAccount
{
    public decimal MinimumBalance { get; }

    public SavingsAccount(string accountNumber, string accountHolder, decimal openingBalance, decimal interestRate, decimal minimumBalance)
        : base(accountNumber, accountHolder, openingBalance, interestRate)
    {
        MinimumBalance = minimumBalance;
    }

    public override void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than zero.");

        decimal resultingBalance = _balance - amount;

        if (resultingBalance < MinimumBalance)
            throw new InvalidOperationException($"Cannot withdraw. Minimum balance of R{MinimumBalance:F2} required.");

        _balance = resultingBalance;
    }
}
